examples/dfem/minimal_surface.cpp

```diff
@@ -0,0 +1,518 @@
+   MFEM_HOST_DEVICE inline
+   static auto coeff(const tensor<T, dim> &a)
+   {
+      return 1_r / sqrt(1_r + sqnorm(a));
+   }
+
+   // Matrix-Free version of the pointwise residual form for the minimal
+   // surface equation.
+   struct MFApply
+   {
+      // Using DifferentiableOperator, we can define the residual form as a
+      // matrix-free operation. This allows us to compute the residual without
+      // explicitly forming any matrices or other large, temporary data
+      // structures.
+      //
+      // The inputs are the gradient of the solution in *reference coordinates*,
+      // the Jacobian of the coordinates, and the integration rule weights.
+      //
+      // The output is the residual in *physical coordinates* which also
+      // includes the necessary transformation from reference to physical
+      // coordinates for the gradient of the test function.
+      //
+      // Due to the description of how this pointwise operation is used in
+      // DifferentiableOperator, we know it is applied to the gradient of the
+      // test function in reference coordinates e.g.
+      // $ \int coeff(\nabla_x u) (\nabla_x u) J^{-T} \det(J) w \nabla_{\xi} v d\xi$
+      MFEM_HOST_DEVICE inline
+      auto operator()(
+         const tensor<dscalar_t, dim> &dudxi,
+         const tensor<real_t, dim, dim> &J,
+         const real_t &w) const
+      {
+         const auto invJ = inv(J);
+         const auto dudx = dudxi * invJ;
+         return tuple{coeff(dudx) * dudx * transpose(invJ) * det(J) * w};
+      }
+   };
+
+   // This is the derivative of the residual form with respect to the
+   // solution $u$.
+   //
+   // The inputs and outputs follow the same rules as the MFApply operator.
+   struct ManualDerivativeApply
+   {
+      MFEM_HOST_DEVICE inline
+      auto operator()(
+         const tensor<real_t, dim> &ddelta_udxi,
+         const tensor<real_t, dim> &dudxi,
+         const tensor<real_t, dim, dim> &J,
+         const real_t &w) const
+      {
+         const auto invJ = inv(J);
+         const auto dudx = dudxi * invJ;
+         const auto ddelta_udx = ddelta_udxi * invJ;
+
+         const auto c = coeff(dudx);
+         const auto term1 = c * ddelta_udx;
+         const auto term2 = c * c * c * dot(dudx, ddelta_udx) * dudx;
+
+         return tuple{(term1 - term2) * transpose(invJ) * det(J) * w};
+      }
+   };
+
+   // This class implements the Jacobian of the minimal surface operator. It
+   // mostly acts as a wrapper to retrieve the Jacobian and apply essential
+   // boundary conditions appropriately.
+   class MinimalSurfaceJacobian : public Operator
+   {
+   public:
+      MinimalSurfaceJacobian(const MinimalSurface *minsurface,
+                             const Vector &x) :
+         Operator(minsurface->Height()),
+         minsurface(minsurface),
+         z(minsurface->Height())
+      {
+         minsurface->u.SetFromTrueDofs(x);
+         auto mesh_nodes = static_cast<ParGridFunction*>
```

Comment:
Specify the type - avoid using auto as the code becomes unreadable.

Reply:
I refer to my comment about `auto` above.